route/option.go

```diff
@@ -232,6 +236,228 @@ func ServiceGroups(serviceGroups []ServiceGroup) Option {
+	// updated information at the vehicle level, directly from the PlanUpdater.
+	func (d fleetData) Update(
+		plan sdkRoute.PartialPlan,
+		vehicles []sdkRoute.PartialVehicle,
+	) (sdkRoute.PlanUpdater, int, bool) {
+		// Deep copy locations stored.
+		locations := make(map[string]int, len(d.Locations))
+		for stopdID, i := range d.Locations {
+			locations[stopdID] = i
+		}
+		d.Locations = locations
+		// Deep copy the data required for the value function.
+		values := make(map[string]int, len(d.vehicleValues))
+		for vehicleID, i := range d.vehicleValues {
+			values[vehicleID] = i
+		}
+		d.vehicleValues = values
+		for _, vehicle := range vehicles {
+			// Update locations based on the changes made on the vehicle state.
+			vehicleID := vehicle.ID()
+			updater := vehicle.Valuer().(vehicleData)
+			for stopdID, i := range updater.Locations {
+				d.Locations[stopdID] = i
+			}
+
+			// Update value function information.
+			value := vehicle.Update()
+			d.fleetValue -= d.vehicleValues[vehicleID]
+			d.vehicleValues[vehicleID] = value
+			d.fleetValue += d.vehicleValues[vehicleID]
+		}
+		// Remove unassigned locations.
+		for _, location := range plan.Unassigned().Slice() {
+			stop := d.stops[location]
+			delete(d.Locations, stop.ID)
+		}
+		return d, d.fleetValue, true
+	}
+
+*/
+func Update(v VehicleUpdater, f PlanUpdater) Option {
+	connect()
+	return updateFunc(v, f)
+}
+
+// FilterWithRoute adds a new VehicleFilter. Compared to the Filter option, the
+// FilterWithRoute option is more flexible. It defines a function that takes an
+// IntDomain of candidate vehicles, an IntDomain of locations that will be
+// assigned to a particular vehicle, and a slice of routes for all vehicles. It
+// returns an IntDomain representing vehicles that cannot service the domain of
+// locations.
+func FilterWithRoute(
+	filter func(
+		vehicleCandidates model.Domain,
+		locations model.Domain,
+		routes [][]int,
+	) model.Domain,
+) Option {
+	connect()
+	return filterWithRouteFunc(filter)
+}
+
+// Sorter sets the given custom vehicle sorter. The vehicle sorter lets you
+// define a function which returns the vehicle indices in a specific order. The
+// underlying engine will try to assign the locations to each vehicle in that
+// returned order.
+func Sorter(
+	sorter func(
+		p PartialPlan,
+		locations model.Domain,
+		vehicles model.Domain,
+	) []int,
+) Option {
+	connect()
+	return sorterFunc(sorter)
+}
+
+// Constraint sets a custom constraint for specified vehicles. It takes two
+// arguments, the constraint to be applied and a list of vehicles, indexed by
+// ID, to which the constraint shall be applied.
+func Constraint(constraint VehicleConstraint, ids []string) Option {
+	connect()
+	return constraintFunc(constraint, ids)
+}
+
+// VehicleConstraint defines an interface that needs to be implemented when
+// creating a custom vehicle constraint.
+type VehicleConstraint interface {
+	Violated(PartialVehicle) (VehicleConstraint, bool)
```

Comment:
For the sake of clarity, can we add a godoc here that states something like `Violated returns true if the vehicle with the current route is not operationally valid`  or something to that extent?

Reply:
And also something that points to the reasoning why you need to return the constraint. Like, when would you alter the constraint...